etc/ha_healthchecker/ha_healthchecker.py

```diff
@@ -0,0 +1,748 @@
+                    LOG.error("Fail Update -- Domain %s from %s to %s" % (
+                        target_domain, TARGET_ORI_BK_IP, TARGET_CHANGE_BK_IP))
+                    LOG.error(
+                        "Details: code-status %s\n         message: %s" % (
+                            res.status_code, res.reason))
+                    return
+        LOG.info("Finish update DNS entry.")
+
+    def _can_start_switch(self):
+        res = []
+        for zk_node in self.zk.list_nodes():
+            if (zk_node.role != 'zookeeper' and
+                    zk_node.switch_status == 'start'):
+                res.append(zk_node.switch_status)
+        return len(set(res)) == 1 and len(res) == 4
+
+    def _is_switching(self):
+        res = []
+        for zk_node in self.zk.list_nodes():
+            if zk_node.role != 'zookeeper':
+                res.append(zk_node.switch_status)
+        return (len(set(res)) == 2 and
+                'start' in set(res) and
+                'end' in set(res))
+
+    def _do_switch(self):
+        if self.node.role == 'master':
+            self._shut_down_all_services(self.node)
+            self.zk.update_node(self.node.name, role='slave', status='down',
+                                switch_status='end')
+            LOG.info("M/S switching: local node, %(role)s node %(name)s is "
+                     "finishd from master to slave. And update it with "
+                     "role=slave and status=down.",
+                     {'role': self.node.role, 'name': self.node.name})
+
+        elif self.node.role == 'slave':
+            if self.node.type == 'zuul':
+                self._change_dns()
+            self.zk.update_node(self.node.name, role='master',
+                                switch_status='end')
+            self._setup_necessary_services_and_check(self.node)
+            LOG.info("M/S switching: local node, %(role)s node %(name)s is "
+                     "finishd from slave to master. And update it with "
+                     "role=master and switch_status=end.",
+                     {'role': self.node.role, 'name': self.node.name})
+
+    def _is_end(self):
+        res = []
+        for zk_node in self.zk.list_nodes():
+            if zk_node.role != 'zookeeper':
+                res.append(zk_node.switch_status)
+        return 'start' not in res
+
+    def run(self):
+        if self.global_analysis():
+            if self._can_start_switch():
+                self._do_switch()
+        if self._is_switching() and self.node.switch_status == 'start':
+            self._do_switch()
+
+        if self._is_end():
+            self.zk.update_node(self.node.name, switch_status=None)
+
+
+class HealthChecker(object):
+    def __init__(self, config_file, heartbeat_timeout):
+        cfg = configparser.ConfigParser()
+        cfg.read(config_file)
+        self.zk = zk.ZooKeeper(cfg)
+        self.heartbeat_timeout = heartbeat_timeout
+
+    def run(self):
+        self.zk.connect()
+
```

Comment:
I think all the check related to `maintaining` status can be moved to here. So that if the node is in `maintaining`, we can skip all the work below.

Reply:
Yeah, but we don't manage node obj in HealthChecker which is just a entry object.. So I think it's good to leave them in the workers objects.